### PR TITLE
Added test1517 for checking ability to receive full HTTP response when POST

### PR DESCRIPTION
**Pull request by Karlson2k**
request is used with slow read callback function.

This test is added to illustrate bug mentioned in issue #657 and PR #668.

### CODE REVIEW

**Comment by Karlson2k** [comment]
Test failure on OSX is unexpected, but reveal problem in libcurl itself, not in test.
If full response is received, libcurl must not report `CURLE_RECV_ERROR`. libcurl should be fixed for OSX.

**Comment by bagder** [comment]
> If full response is received, libcurl must not report CURLE_RECV_ERROR. libcurl should be fixed for OSX

I'm not convinced. Where exactly is the OS X error you say?

What does libcurl return on Windows for this failure?

**Comment by Karlson2k** [comment]
OSX error on Travis automated testing.
libcurl returns error 56, same as on Windows.
But on OSX libcurl get full HTTP server response, while on Windows libcurl got empty response.

**Comment by bagder** [comment]
I suspect it is the other way around. The fact that libcurl returns CURLE_RECV_ERROR on OS X means that recv() actually returned an error (there are a few other possibilities too but I think they're less likely). After all, the Linux version and the OS X version run the same code.

**Comment by Karlson2k** [comment]
(updated test data with more comments)
@bagder Look at test and test results. Actually libcurl **correctly** sent request and **correctly** received reply. May be OSX report error for sending last request block (request body) or for some other reason. If OSX reported error then error must be analyzed or libcurl logic must be changed as replying right after request headers does not violate HTTP standard.
In other words: something must be changed (for OSX) so libcurl will not report **error** after **correct** receive of full HTTP response.
In case of Windows libcurl **do not receive** HTTP response at all, so workarounds like #668 should be implemented.

**Comment by bagder** [comment]
I believe you're relying on behavior that is both timing dependent and depending on implementation details that isn't specified in any docs. The data sent by the server before it closes the socket is not guaranteed to reach the other end.

[The ultimate SO_LINGER page](https://www.archive.org/3/items/TheUltimateSo_lingerPageOrWhyIsMyTcpNotReliable/the-ultimate-so_linger-page-or-why-is-my-tcp-not-reliable.html) says (and I quote the above mentioned page here):

Let’s say we run the following two programs on two POSIX compliant operating systems, with the intention of sending 1 million bytes from program A to program B (programs can be found here):

A:
   
      sock = socket(AF_INET, SOCK_STREAM, 0);  
      connect(sock, &remote, sizeof(remote));
      write(sock, buffer, 1000000);             // returns 1000000
      close(sock);

B:

    int sock = socket(AF_INET, SOCK_STREAM, 0);
    bind(sock, &local, sizeof(local));
    listen(sock, 128);
    int client=accept(sock, &local, locallen);
    write(client, "220 Welcome\r\n", 13);

    int bytesRead=0, res;
    for(;;) {
        res = read(client, buffer, 4096);
        if(res < 0)  {
            perror("read");
            exit(1);
        }
        if(!res)
            break;
        bytesRead += res;
    }
    printf("%d\n", bytesRead);

Quiz question - what will program B print on completion?

A) 1000000
B) something less than 1000000
C) it will exit reporting an error
D) could be any of the above

The right answer, sadly, is ‘D’.

**Comment by Karlson2k** [comment]
Of course, this is a nature of networks.
But you didn't check which error returned by system. If it's send error, this doesn't mean that receive was incorrect.
Moreover, you have some tips: protocol is HTTP 1.1, end of header is received (empty line with "\r\n"), connection is "close", no Content-Length is specified and no chunked encoding, so this should be header-only response.
Anyway, some kinds of *errors* are not considered as real *errors* already (`EAGAIN` or `EWOULDBLOCK`). So may be libcurl just need to check which error is reported by OSX in that case.

Test itself is correct: HTTP server allowed to reply and close connection, result does not depend on timing: response is **always** received on OSX during this test.

**Comment by Karlson2k** [comment]
Just to be clear: the **only** problem on OSX is final error code, reported by libcurl: it's 56 (`CURLE_RECV_ERROR`).
Test itself on OSX is actually passed, just like on Linux: server **response** is fully and correctly **received** by libcurl. But on Linux libcurl returns 0 while on OSX on the same situation libcurl returns 56.

Look on test logs: client log has server response.

**Comment by bagder** [comment]
> connection is "close", no Content-Length is specified and no chunked encoding, so this should be header-only response.

No, that means the closing of the socket signals the end of data. A 500 response (which you made test 1517 respond with) includes a response-body.

> on OSX on the same situation libcurl returns 56

Yes, quite clearly. The test case as currently written is unreliable.

**Comment by Karlson2k** [comment]
@bagder You are right. While HTTP 1.1 server must not send responses without both "Content-Length:" and "Transfer-Encoding: chunked", client should accept such responses.
I updated testcase, so it should be correct now.

**Comment by Karlson2k** [comment]
With updated testcase libcurl passes test even on buggy OSX.
I think this is a real bug of OSX because recv() returned `-1` while there were no receive errors, but libcurl already correctly handle this situation.

**Comment by Karlson2k** [comment]
On Windows platform bug is present in all checked versions: WinXP, Vista, 7, 8, 8.1, 8.1u1 and 10.

**Comment by Karlson2k** [comment]
Updated with proper commit description.
